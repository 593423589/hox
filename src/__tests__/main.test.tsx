import { setModel, useModel, selectModel, withModel } from "..";
import * as React from "react";
import { Component, FC, memo, useState } from "react";
import * as testing from "@testing-library/react";

test("simple", function() {
  function useCounter() {
    const [count, setCount] = useState(0);
    const decrement = () => setCount(count - 1);
    const increment = () => setCount(count + 1);
    return { count, decrement, increment };
  }

  setModel("counter", useCounter);

  const App: FC = () => {
    const counter = useModel<typeof useCounter>("counter");

    return (
      <div>
        <button onClick={counter.increment}>Change</button>
        {counter.count}
      </div>
    );
  };
  const renderer = testing.render(<App />);
  expect(renderer.asFragment()).toMatchSnapshot();
  testing.fireEvent.click(testing.getByText(renderer.container, "Change"));
  expect(renderer.asFragment()).toMatchSnapshot();
});

test("withModel", function() {
  function useCounter() {
    const [count, setCount] = useState(0);
    const decrement = () => setCount(count - 1);
    const increment = () => setCount(count + 1);
    return { count, decrement, increment };
  }

  setModel("counter", useCounter);

  interface Props {
    counter: ReturnType<typeof useCounter>;
  }

  class App extends Component<Props> {
    render() {
      return (
        <div>
          <button onClick={this.props.counter.increment}>Change</button>
          {this.props.counter.count}
        </div>
      );
    }
  }

  type Counter = ReturnType<typeof useCounter>;

  const AppWithModel = withModel("counter", (model: { counter: Counter }) => ({
    counter: model.counter
  }))(App);
  const renderer = testing.render(<AppWithModel />);
  expect(renderer.asFragment()).toMatchSnapshot();
  testing.fireEvent.click(testing.getByText(renderer.container, "Change"));
  expect(renderer.asFragment()).toMatchSnapshot();
});

test("withModel without mapModelToProps", function() {
  function useCounter() {
    const [count, setCount] = useState(0);
    const decrement = () => setCount(count - 1);
    const increment = () => setCount(count + 1);
    return { count, decrement, increment };
  }

  setModel("counter", useCounter);

  interface Props {
    model: {
      counter: ReturnType<typeof useCounter>;
    };
  }

  class App extends Component<Props> {
    render() {
      return (
        <div>
          <button onClick={this.props.model.counter.increment}>Change</button>
          {this.props.model.counter.count}
        </div>
      );
    }
  }

  type Counter = ReturnType<typeof useCounter>;

  const AppWithModel = withModel("counter")(App);
  const renderer = testing.render(<AppWithModel />);
  expect(renderer.asFragment()).toMatchSnapshot();
  testing.fireEvent.click(testing.getByText(renderer.container, "Change"));
  expect(renderer.asFragment()).toMatchSnapshot();
});
