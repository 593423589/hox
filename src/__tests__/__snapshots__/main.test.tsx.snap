// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple 1`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    0
  </div>
</DocumentFragment>
`;

exports[`simple 2`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    1
  </div>
</DocumentFragment>
`;

exports[`withModel 1`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    0
  </div>
</DocumentFragment>
`;

exports[`withModel 2`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    1
  </div>
</DocumentFragment>
`;

exports[`withModel without mapModelToProps 1`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    0
  </div>
</DocumentFragment>
`;

exports[`withModel without mapModelToProps 2`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    1
  </div>
</DocumentFragment>
`;
